### ISSUE TITLE
ATTime treats timezones differently for absolute vs. relative times

### ISSUE THREAD

**Issue by MFAnderson** [graphite-project/graphite-web#1151]
The project I'm working on is trying to do a version bump on graphite (to 0.9.13-pre1) to obtain some fixes that we need, but we've run into a problem with some of the time ranges that we are using. Specifically, we have graphs using parameters like `from=midnight&to=midnight_tomorrow&tz=America/New_York`, which used to give data for the full current day defined relative to Eastern time but now gives a range starting at 7PM the previous day. I've done some digging and identified render/attime.py as the source of the change. 

When parseATTime is called with an absolute time and a timezone, the datetime object returned is for the specified time in the specified timezone. If however a relative time is given such as "midnight", "noon", "midnight_yesterday", etc. the returned datetime is for the evaluated time in UTC converted to the specified timezone.

``` python
parseATTime("midnight", pytz.timezone("America/New_York"))
=> datetime.datetime(2015, 2, 11, 19, 0, 0, 570298, 
                tzinfo=<DstTzInfo 'America/New_York' EST-1 day, 19:00:00 STD>)

parseATTime("12:00_20150211", pytz.timezone("America/New_York"))
=> datetime.datetime(2015, 2, 11, 12, 0, tzinfo=<DstTzInfo 'America/New_York' EST-1 day, 19:00:00 STD>)
```

This change in behaviour from 0.9.12 that appears to have been introduced in the process of fixing some other time issues in commits made by @brutasse to graphite-api and ported to graphite-web by @deniszh , but I'm unclear if this is an intended change in behaviour or a regression. If it is a regression, I would be happy to either work with an existing contributor or submit a pull request with a fix. Anyone have any insight?

**Comment by brutasse** [comment]
That's not intended behavior. The issue is here:

https://github.com/graphite-project/graphite-web/blob/master/webapp/graphite/render/attime.py#L68

`timezone.now()` takes the project timezone regardless of what was passed. The timezone should be passed all the way to `parseTimeReference` instead.

**Comment by MFAnderson** [comment]
Yeah, I had come to more or less the same conclusion. My one concern regarding passing timezone through to `parseTimeReference` is that I think we'll run into issues around DST again once the timedelta from `parseOffset` gets added. The solution I was thinking of was to keep `parseTimeReference` dealing with naive datetimes and only localize after the offset has been included. From what I can gather reading the pytz docs this *should* avoid the weird DST issues that create the need for `normalize`. I've got an attempt at a fix for this on my fork MFAnderson/graphite-web@86999666073569a0ed0568654a3fdffca92f202f but I want to add some test coverage around the DST specific issues before submitting a pull request. Can you take a look at that code and see if it makes sense to you?

**Comment by MFAnderson** [comment]
I've been playing around with stuff on the DST boundary and realized there's a couple of weird things  that don't necessarily follow intuitive assumptions. Namely, the way `localize` works means that for, say, "America/New_York" 
```python
tz.localize(datetime(2015, 3, 8, 2, 0)) == tz.localize(datetime(2015, 3, 8, 3, 0))
=> True
```
because it treats the first datetime as being in EST and the second in EDT.

as it pertains to attime, this results in 
```python
parseATTime("02:00_20150308", tz) == parseATTime("03:00_20150308", tz)
=> True
```
and similarly with the fix I tried out
```python
#given the current day is the day DST begins
parseATTime("midnight+2h", tz) == parseATTime("midnight+3h", tz)
=> True
```

Other than this bit of weirdness, I've confirmed that the fix at least ensures that absolute and relative times behave consistently with regard to timezone.

**Comment by urtzurd** [comment]
Hi,

We're in the same situation trying to bump the version to start using some of the new functionalities. We have been testing the latest development version and we have noticed some very strange behaviours with relative time ranges. I'm not sure if they might be directly related to the latests changes introduced by this merge.

Our scenario is as following: we have our graphite server configured to use UTC timezone and so is configured graphite-web too. We set the timezone when creating graphs/dashboards. Using the latest version we have experienced that when we have a graph to show the past 24h and the timezone is Europe/London, is shows 24h of data, but not from the current time, but with 1h difference. So if it's 9:00AM London time, it will show 24h data until 8:00AM London time. I have tried with Europe/Madrid and it has a similar problem (it shows until 9:00AM when it should show until 11:00AM).

May I be missing some new configuration variable or is there a new bug?
Best regards.

**Comment by deniszh** [comment]
@urtzurd:
Looks like same issue,  but it was fixed in master and 0.9.x branch from end of February.
Which version are you using - latest master from github?

**Comment by urtzurd** [comment]
Yes, we are testing master. I guess that master will be the branch that will become 0.10 in the future and I should stick to the 0.9.x in the meantime. Will give it a try just now. Thanks a lot for the very quick response!

**Comment by urtzurd** [comment]
I've been testing 0.9.x and definitely there's still some issue with the timezones. It seems that it's dashboard related only. I've plenty of graphs saved with Europe/London timezone and they display properly when loaded. The problem is when those graphs are included into a dashboard, sometimes they work, sometimes not. The behaviour is very erratic and sometimes the data appears correctly with all the data until now, then sometimes after a refresh it goes wrong and moves back not displaying the latest data.

I tried creating a dashboard from scratch and it showed the same issues. First I thought that it could be related to format changes in the dashboard JSON, so I tried cleaning up them and removed repeated parameters like from and until inside each graph, but that didn't make any difference.

As a workaround we have changed all our dashboards to be UTC and that seems to work.
